server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java

```diff
@@ -50,26 +50,37 @@ public TransferManager(final BlobContainer blobContainer, final ExecutorService
     /**
      * Given a blobFetchRequest, return it's corresponding IndexInput.
      * @param blobFetchRequest to fetch
-     * @return future of IndexInput augmented with internal caching maintenance tasks
+     * @return The IndexInput of the requested data
      */
-    public CompletableFuture<IndexInput> asyncFetchBlob(BlobFetchRequest blobFetchRequest) {
-        return asyncFetchBlob(blobFetchRequest.getFilePath(), () -> {
+    public IndexInput fetchBlob(BlobFetchRequest blobFetchRequest) {
+        final CompletableFuture<IndexInput> originFuture = invocationLinearizer.linearize(blobFetchRequest.getFilePath(), p -> {
             try {
-                return fetchBlob(blobFetchRequest);
+                return fetchOriginBlob(blobFetchRequest);
             } catch (IOException e) {
-                throw new IllegalStateException(e);
+                throw new UncheckedIOException(e);
             }
         });
-    }
 
-    private CompletableFuture<IndexInput> asyncFetchBlob(Path path, Supplier<IndexInput> indexInputSupplier) {
-        return invocationLinearizer.linearize(path, p -> indexInputSupplier.get());
+        try {
+            final IndexInput origin = originFuture.get();
+            // The origin instances stays in the cache with a ref count of zero
+            // and must be cloned before being returned.
+            return origin.clone();
```

Comment:
@andrross I am struggling to understand this change: `fetchBlob` used to return cloned version but now there is a blocking call and, well, cloning. The only effect I could deduct from that is that cloning would happen in different threads, was that the issue?

Reply:
The crux of the issue is that `invocationLinearizer.linearize()` can result in concurrent calls being returned the _same_ instance, so that means the clone call must after the `linearize()` call. Each concurrent thread must independently clone the result, and that wasn't happening before.

I did move the blocking call from `OnDemandBlockSnapshotIndexInput` to here. That wasn't strictly necessary, but the clone call must happen after redeeming the future and I wanted to keep all the cloning logic inside this class.

Reply:
@reta The sequence was like this:

1. download blob
1. clone
1. wrap in future and return (potentially to multiple threads)
1. get result from future

It has changed to this:


1. download blob
1. wrap in future and return (potentially to multiple threads)
1. get result from future
1. clone

Reply:
@reta That was my first instinct but I think each concurrent thread is given the same CompletableFuture instance, so for `n` threads only one clone would be made and shared across threads.

Reply:
@reta What I could have done is just move the clone call from `TransferManager::fetchBlob` to `OnDemandBlockSnapshotIndexInput::fetchBlob`, but I felt that TransferManager is the one that should be keeping track of what is cloned and should not be returning things that are not safe to use directly, which led to a slightly larger refactoring.

Reply:
Oh I see what you mean, no I think the composition should be fine:
```
return asyncFetchBlob(blobFetchRequest.getFilePath(), () -> {
            try {
                return fetchBlob(blobFetchRequest);
            } catch (IOException e) {
                throw new IllegalStateException(e);
            }
        }).thenApply(r -> r.clone());
```

Indeed, `fetchBlob` would return a value from first concurrent execution (if any), but each invoker would get an unique clone of the value.

Reply:
@andrross I agree that `TransferManager` should keep these guarantees

Reply:
> I think the composition should be fine

You're right, `thenApply` will do exactly what I want since each invocation will create a new future instance with the `thenApply` function applied to it.